data/common.yaml

```diff
@@ -14,6 +14,102 @@ gpfs::add_client::ssh_private_key_path: "/root/.ssh/id_gpfs"
 gpfs::add_client::ssh_public_key_type: "rsa"
 gpfs::add_client::script_tgt_fn: "/root/gpfs_add_client.sh"
 gpfs::cron::accept_license: false
+gpfs::health::enabled: false
+gpfs::health::file_base_name: "gpfs_client_health"
+gpfs::health::telegraf_cfg:
+  data_format: "influx"
+  interval: "1m"
+  timeout: "30s"
+gpfs::health::telegraf_script: |2+
```

Comment:
Can we define this script in a file instead of in hiera?

It doesn't seem like the script will need to be customized per cluster too often

Reply:
I think at least the interval and timeout need to be configurable via hiera.

I initially had a much simpler version of this, but ended up copying this functionality directly from the lustre client health puppet module: https://github.com/ncsa/puppet-profile_lustre/blob/main/manifests/telegraf/lustre_client_health.pp
I think it is good to keep how they deal with telegraf being the same.

Reply:
Sorry, I don't like how github highlights more lines than what I'm trying to. I was asking about the script in line 23 for `gpfs::health::telegraf_script`

Reply:
Personally I'm a fan of putting scripts in hiera when we can. It provides more flexibility just in case we need it. We already use this approach for some other scripts across various puppet profiles.

I realize that it adds a bit of clutter to the hiera data and makes it harder to view, but it really only clutters up the puppet profile's hiera data (as opposed to a project's control repo).

If you have an argument against that approach, I'd like to hear it.

Reply:
I see the flexibility argument but I thinks it's too flexible.

Leaving the whole script in hiera means any cluster can make any change they want to the script, which then causes competing versions of the script.

If we expect the script to need some customization it's best if it's in a well defined way, like making it source a config file that lets it set some control variables to change behavior. Otherwise any cluster is free to change it in any way, even ones which will make it incompatible with other versions of the script.

Later if we make an improvement to the script, it's not as simple as just checking the version of the module across our clusters, we also have to check if `gpfs::health::telegraf_script` is getting overridden in each cluster. If we do find a cluster that overrides the script we have to manually refactor our new improvement changes into whatever changes the cluster made previously (and there is no guarantee that the previous deviations the cluster made won't make it incompatible with the improvement we are now trying to add).

Reply:
I feel like keeping scripts in separate files fits better with how my brain thinks. I can really only speak on the Telegraf side (and not all the other things you all are doing with puppet) but I try really hard to avoid (and I think I've done so successfully) there being anything dynamic (per cluster/deployment scenario) in these custom exec scripts. Anything custom is source'd from a config file I believe is already in hiera which that part makes sense to me, and if anything needed to be dynamic in the script itself I'd just tack another variable into the config file.

When doing development I usually work outside of any repo, for example if adding a feature to a slurm check I go to a scheduler and make a copy of the current running script, and then edit that copy (since Telegraf won't be trying to execute that copied-aside version). I then execute it by hand when I want to see how it does/is it working as expected as I'm writing the new feature (or patching a bug). Once I have the new working/tested version of the check, my current workflow is to then copy the script in it's entirety, open the script file in the puppet repo, nuke the entire contents of that file, then paste in the new version. That way I know things match 100% between what I tested as working and what I'm asking folks to review in git/push to production. With the script in hiera, I think I could technically do most of that, but I'd have to make sure I just nuke the script part of hiera and not anything else, which I agree most of the time isn't hard but I'd be more worried about accidentally deleting things I shouldn't.

Reply:
Sounds good. I'll change it to a static file.

fyi: I'm pretty sure the lustre client health check script is currently a template file in Puppet. I didn't look at the details, but presumed something must be dynamic in that script.

Reply:
The script is now a static file. 

The script here is slightly different than the one at https://git.ncsa.illinois.edu/ici-monitoring/ici-developed-checks/-/blob/main/spectrum_scale_client/ss_health.sh . At some point we should update the version in that git repo to match.